File: 豆豆/2021-12-17/Decoded String at Index.cpp
//
//  main.cpp
//  Test
//
//  Created by 洪立恒 on 2021/12/17.
//
#include <cstring>
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    string decodeAtIndex(string s, int k) {
        string temp = "";
        string stemp = "";
        char Ans[k];
        for(int i = 0 ; i<s.length();i++){
            if(k==1){
                stemp = s[0];
                return stemp;
            }
            if(isdigit(s[i])){
                for(int a = 0;a<s[i]-'0';a++){
                    stemp += temp;
                }
                temp = stemp;
                stemp = "";
            }else temp += s[i];
            if(temp.length()>=k){
                for(int j = k-1;j<=k;j++){
                    Ans[j] = temp[j];
                }
            }
        };
        temp = Ans[k-1];
        return temp;
    };
};

int main(){
    string ss = "leet2code3";
    int kk = 10;
    Solution Ans;
    cout<< Ans.decodeAtIndex(ss, kk)<<endl;
    
};

File: 許靜玟/2021.12.17/DecodedStringatIndex.py
class Solution:
    def decodeAtIndex(self, s: str, k: int) -> str:
        idx, ans = 0, ""
        for a in range(len(s)):
            if idx >= k: return ans[k-1]

            if s[a].isdigit():
                idx *= int(s[a])
                ans *= int(s[a])
            else:
                idx += 1
                ans += s[a]
        return ans[k-1]

# class Solution:
#     def decodeAtIndex(self, S, K):   
#         A = [1]
#         for x in S[1:]:
#             if A[-1] >= K : break

#             if x.isdigit():
#                 A.append( A[-1]*int(x) )
#             else:
#                 A.append( A[-1]+1 )
        
#         for i in reversed(range(len(A))):
#             K %= A[i]
#             if not K and S[i].isalpha():
#                 return S[i]